File: leetcode/547-friend-circles/main.go
package main

import "fmt"

/*
	1st approach: union find
	- put the friends connections into an union find
	- return the total 'cluster' as a result

	Time		O(nlogn)
	Space		O(n)
	28 ms, faster than 100.00%
*/
func findCircleNum(M [][]int) int {
	if len(M) == 0 || len(M[0]) == 0 {
		return 0
	}
	uf := Constructor(len(M))
	for i := 0; i < len(M); i++ {
		for j := i; j < len(M); j++ {
			if M[i][j] == 1 {
				uf.Union(i, j)
			}
		}
	}
	return uf.GetCount()
}

type UnionFind struct {
	Count int
	Ids   []int
	Caps  []int
}

func Constructor(n int) UnionFind {
	ids := []int{}
	caps := []int{}
	for i := 0; i < n; i++ {
		ids = append(ids, i)
		caps = append(caps, 1)
	}
	return UnionFind{n, ids, caps}
}

func (this *UnionFind) GetCount() int {
	return this.Count
}

// O(logN)
func (this *UnionFind) Find(key int) int {
	if key < len(this.Ids) {
		// loop to find to ultimate root
		cur := key
		for cur != this.Ids[cur] {
			cur = this.Ids[cur]
		}
		return cur
	}
	return -1
}

// union to the bigger tree, O(N)
func (this *UnionFind) Union(p int, q int) {
	if p < 0 || p+1 > len(this.Ids) || q < 0 || q+1 > len(this.Ids) {
		return
	}
	pId := this.Find(p)
	qId := this.Find(q)
	if pId == qId {
		return
	}

	if this.Caps[pId] < this.Caps[qId] {
		this.Ids[pId] = qId
		this.Caps[qId] += this.Caps[pId]
	} else {
		this.Ids[qId] = pId
		this.Caps[pId] += this.Caps[qId]
	}
	this.Count--
}

func main() {
	a := [][]int{
		{1, 1, 0},
		{1, 1, 0},
		{0, 0, 1},
	}
	fmt.Println(findCircleNum(a))

	a = [][]int{
		{1, 1, 0},
		{1, 1, 1},
		{0, 1, 1},
	}
	fmt.Println(findCircleNum(a))

	a = [][]int{
		{1, 1, 0, 0},
		{1, 1, 0, 1},
		{0, 0, 1, 0},
		{0, 1, 0, 1},
	}
	fmt.Println(findCircleNum(a))
}

File: leetcode/547-friend-circles/main.py
"""
    1st approach: union find
    - put the friends connections into an union find
    - return the total 'cluster' as a result

    Time		O(NNlogN) -> O(N^3)
    Space		O(N)
    204 ms, faster than 24.89
"""


class Solution(object):
    def findCircleNum(self, M):
        """
        :type M: List[List[int]]
        :rtype: int
        """
        N = len(M)
        uf = UnionFind(N)
        for i in range(N):
            for j in range(i+1, N):
                if M[i][j] == 1:
                    uf.union(i, j)
        return uf.count


class UnionFind(object):
    def __init__(self, n):
        self.count = n
        self.ids = {}
        self.caps = {}
        for i in range(n):
            self.ids[i] = i
            self.caps[i] = 1

    def find(self, key):
        # loop to find to ultimate root
        if key not in self.ids:
            return None
        cur = key
        while cur != self.ids[cur]:
            cur = self.ids[cur]
        return cur

    def union(self, p, q):
        if p not in self.ids or q not in self.ids:
            return

        pId = self.find(p)
        qId = self.find(q)
        if pId == qId:
            return
        # attach to the larger tree
        if self.caps[pId] < self.caps[qId]:
            self.ids[pId] = qId
            self.caps[qId] += self.caps[pId]
        else:
            self.ids[qId] = pId
            self.caps[pId] += self.caps[qId]
        self.count -= 1


"""
    2nd approach: bfs
    - get the friend list for each friend
    - for each person, bfs to find all of his friends and put them into history
    - after finished one bfs, it means that we are done with one 'cluster'

    Time		O(NN)
    Space		O(N)
    220 ms, faster than 23.64%
"""


class Solution(object):
    def findCircleNum(self, M):
        n = len(M)
        graph = defaultdict(set)
        for i in range(n):
            for j in range(n):
                if M[i][j] == 1:
                    graph[i].add(j)
                    graph[j].add(i)
        res = 0
        seen = set()
        for i in range(n):
            if i in seen:
                continue
            self.bfs(graph, i, seen)
            res += 1
        return res

    def bfs(self, graph, start, seen):
        q = [start]
        while len(q) > 0:
            node = q.pop(0)
            if node in seen:
                continue
            seen.add(node)
            if node not in graph:
                continue
            for nb in graph[node]:
                q.append(nb)


a = [[1, 1, 0],
     [1, 1, 1],
     [0, 1, 1]]
print(Solution().findCircleNum(a))

a = [[1, 1, 0],
     [1, 1, 0],
     [0, 0, 1]]
print(Solution().findCircleNum(a))

a = [[1, 1, 0, 0], [1, 1, 0, 1], [0, 0, 1, 0], [0, 1, 0, 1]]
print(Solution().findCircleNum(a))
